Sources/CollectionDiffing.swift

```diff
@@ -0,0 +1,560 @@
+		return lift { $0.diff(identifyingBy: identifyingStrategy, comparingBy: comparingStrategy) }
+	}
+}
+
+extension PropertyProtocol where Value: Collection, Value.Indices.Iterator.Element == Value.Index {
+	/// Compute the difference of `self` with regard to `old` by value equality.
+	///
+	/// `diff(with:)` works best with collections that contain unique values.
+	///
+	/// If the elements are repeated per the definition of `Element.==`, `diff(with:)`
+	/// cannot guarantee a deterministic stable order, so these would all be uniformly
+	/// treated as removals and inserts.
+	///
+	/// - precondition: The collection type must exhibit array semantics.
+	///
+	/// - parameters:
+	///   - keyGenerator: A unique identifier generator to apply on elements.
+	///   - comparator: A comparator that evaluates two elements for equality.
+	///
+	/// - complexity: O(n) time and space.
+	public func diff<Identifier: Hashable>(identifier: @escaping (Value.Iterator.Element) -> Identifier, areEqual: @escaping (Value.Iterator.Element, Value.Iterator.Element) -> Bool) -> Property<Snapshot<Value, Changeset>> {
+		return lift { $0.diff(identifier: identifier, areEqual: areEqual) }
+	}
+}
+
+// MARK: - Implementation details
+
+// The key equality implies only referential equality. But the value equality of the
+// uniquely identified element across snapshots is uncertain. It is pretty common to diff
+// elements with constant unique identifiers but changing contents. For example, we may
+// have an array of `Conversation`s, identified by the backend ID, that is constantly
+// updated with the latest messages pushed from the backend. So our diffing algorithm
+// must have an additional mean to test elements for value equality.
+
+private final class DiffEntry {
+	var occurenceInOld = 0
+	var occurenceInNew = 0
+	var locationInOld: Int?
+}
+
+private enum DiffReference {
+	case remote(Int)
+	case table(DiffEntry)
+}
+
+// FIXME: Swift 4 Associated type constraints
+// extension Collection {
+extension Collection where Index == Indices.Iterator.Element {
+	// @testable
+	internal static func diff<Identifier: Hashable>(
+		previous: Self,
+		current: Self,
+		identifier: (Iterator.Element) -> Identifier,
+		areEqual: (Iterator.Element, Iterator.Element) -> Bool
+	) -> Changeset {
+		switch Self.self {
+		case is Array<Iterator.Element>.Type:
+			fallthrough
+		case is ContiguousArray<Iterator.Element>.Type:
+			// The standard library does not copy the content when `ContiguousArray.init`
+			// is passed an `Array` backed by a native buffer or a `ContiguousArray`. So
+			// we exploit this optimization to obtain an unsafe buffer pointer.
+			//
+			// `ArraySlice` is not supported at the moment, since it may have a non-zero
+			// start index that needs to be taken care of.
+			return ContiguousArray(previous).withUnsafeBufferPointer { previousBuffer in
```

Comment:
I don't love this. 🙈 

Have you measure this optimization? I wouldn't think that this would take up a significant portion of the runtime.